docs/src/appendix/packaging.md

```diff
@@ -0,0 +1,501 @@
+16. Write 32 zeroes to the output buffer
+17. Append the header bytes
+18. Compute the HMAC-SHA256 signature of the content buffer, using the package
+name and version concatenated together and hex-encoded as the signing key, and append the signature to the output buffer
+19. Append the content bytes
+20. Compute the HMAC-SHA256 signature of the output buffer, sans the first 40 bytes, using the header magic, schema version, and content signature concatenated together (hex-encoded) as the signing key.
+21. Write the signature we just computed to the 32-byte region of step 16
+
+The content of the output buffer can be described using the following structure (pseudo-Rust):
+
+```rust
+#[repr(C)]
+pub struct Package {
+    magic: [u8; 8],            // b"MIDENPF\0"
+    digest: [u8; 32],
+    header: Header,
+    content_digest: [u8; 32],
+    content: [u8], // dynamically-sized trailing bytes
+}
+
+#[repr(C)]
+pub struct Header {
+    schema_version: u8,        // 0x1
+    ty: PackageType,           // 1 byte
+    features: Features,        // 1 byte
+    name_len: usize,           // LEB128-encoded
+    version_len: usize,        // LEB128-encoded
+    miden_version_len: usize,  // LEB128-encoded
+    start: [u8; 32],
+    num_segments: usize,       // LEB128-encoded
+    segments: [SegmentInfo; num_segments],
+    num_dependencies: usize,   // LEB128-encoded
+    dependencies: [DependencyInfo; num_dependencies],
+}
+
+#[repr(u8)]
+pub enum PackageType {
+    Library,
+    Program,
+}
+
+bitflags::bitflags! {
+    pub struct Features: u8 {
+        /// This package provides documentation
+        const DOCS = 0x01;
+        /// This package provides debug information
+        const DEBUG = 0x02;
+    }
+}
+
+#[repr(C)]
+pub struct SegmentInfo {
+    /// The relative offset in the segment data section of the package content
+    /// where the data for this segment can be found
+    offset: usize,
+    /// The offset in words where the segment should be placed
+    base: u32,
+    /// The size in bytes of the raw binary data
+    size: u32,
+}
+
+#[repr(C)]
+pub struct DependencyInfo {
+    /// The relative offset in the segment data section of the package content
+    /// where the data for this segment can be found
+    offset: usize,
+    /// Flags used to indicate how this dependency is used
+    flags: u8,          // vendored=0x01
+    name_len: usize,    // LEB128-encoded
+    version_len: usize, // LEB128-encoded
+    digest_len: usize,  // LEB128-encoded
+}
+```
+
+Without the header, the content of the package is an opaque bag of bytes, but the header allows interpreting the content as if it had the following structure (pseudo-Rust):
+
+```rust
+#[repr(C)]
+pub struct Content {
+    name: [u8; name_len],
+    version: [u8; version_len],
+    miden_version: [u8; miden_version_len],
+    segment_data: [SegmentData; num_segments],
+    dependency_infos: [DependencyData; num_dependencies],
+    wit: Component,
+    mast: MastForest,
```

Comment:
Should we consider compressing the binary data for `Component` and `MastForest`? It seems like data savings might be noticable. We could make the compression optional and hide it behind a flag(s) in `features` bitflags.

Reply:
Both of those structures have compact binary representations, so they should already be quite compact. That said, my intention is that the package format is purely a container, so compression can be applied to the container as a whole, using whatever compression algorithm is best suited to the situation.

This would be akin to tarballs, which are uncompressed by default, but can be compressed using a variety of different schemes (bzip2, gunzip, lzma, etc.), yet understood by a single tool (i.e. `tar`) simply using the file extension as a signal for which compression scheme was used. We don't necessarily need to go that far ourselves, but I think the idea would be to separate compression from the container format itself, so that a choice can be made for whether to compress, and how to do so, on a case-by-case basis.

Reply:
I think this is somewhat related to the comment I made about off-chain vs. on-chain formats. I think for off-chain format, general compression algorithm would probably work fine. But for on-chain format (when we don't really care much about deserialization speed), we should be able to do much better than a generic algorithm would because we can omit serialization of all intermediate hashes.

Reply:
If it goes on-chain, we should squeeze every byte we can.

Reply:
@bobbinth Can you elaborate on how we can omit serialization of the full MAST (at least the parts that are "new", i.e. do not reference externally-defined code)? My understanding is that having the MAST root is not equivalent to having the actual MAST, just that all you need is the MAST root to validate the entire tree. The package format is intended to solve the question of obtaining the tree in the first place.

That said, I think we can certainly use your own homegrown schemes for encoding to binary, but it would be hard to beat off-the-shelf schemes for packages, since the package itself contains a variety of different bits of data, some of which we don't control (i.e. rodata segments).

Reply:
I wrote a part of the answer in https://github.com/0xPolygonMiden/compiler/pull/132#pullrequestreview-1999498527 - but to summarize: if our goal is to compress the binary as much as possible, there are a couple of things we can do differently:

1. We can omit serializing all digests and instead serialize only basic blocks (i.e., span nodes) and some minimal metadata for all other nodes (i.e., on the order on a few bits per node). This would effectively reduce the size of the binary by close to $32 \cdot n$ bytes where $n$ is the number of nodes in the MAST. It will also make deserialization much more computationally intensive - so, there are tradeoffs.
2. We can make the format more "serial" - i.e., get rid of ability to do arbitrary lookups into the serialized MAST. This should allow shaving off some bytes too, but the impact is more difficult to estimate.